### ISSUE TITLE
AutoTimer editors incorrectly handle "Only match between dates"

### ISSUE THREAD

**Issue by Grumpy-Geoff** [E2OpenPlugins/e2openplugin-OpenWebif#1405]
If an AutoTimer definition is date-restricted then it contains the 'after="date#1"' and 'before="date#2"' conditions.
e.g.
```
root@beyonwizv2:~# grep Rosehaven /etc/enigma2/autotimer.xml
 <timer name="Rosehaven" match="Rosehaven" enabled="yes" id="82" after="1632931200" before="1633017600" avoidDuplicateDescription="2" searchType="exact" searchCase="sensitive" overrideAlternatives="1">
where the after condition value evaluates to (Perth) -
root@beyonwizv2:~# date --date=@1632931200    ## after
Thu Sep 30 00:00:00 AWST 2021
root@beyonwizv2:~# date --date=@1633017600    ## before
Fri Oct  1 00:00:00 AWST 2021
root@beyonwizv2:~#
```
The AutoTimer GUI editor shows for that definition -
_"Not before - 30 Sep 2021
Not after - 01 Oct 2021"_

With regard to the "after/before" (displayed - not before/not after) dates above, the OWIF Classic and Modern editors have different problems with the "Only match between dates" handling.

Classic editor:
In the above case the Classic editor displays these dates incorrectly reversed as -
_"Only match between dates:  **To** date: 30.09.2021  **From** date: 01.10.2021"_
If the date is changed, it will usually fail with an error as the OWIF editor treats them as labeled.

Modern editor:
The modern editor handles this a bit better, but still incorrectly, as it displays the dates reduced by one day each -
_"**From** date: 29/09/2021  **To** date: 30/09/2021"_
But if a date is changed then we can see the time has actually been reduced by 16 hours (57,600 secs).
e.g was -
after="1632931200" before="1633017600" 
Then after extending the "To date" (only) by a day in the modern editor and saving, it reduces the dates/times by 16 hours -
```
<timer name="Rosehaven" match="Rosehaven" enabled="yes" id="82" after="1632873600" before="1633046400" offset="5,10" avoidDuplicateDescription="2" searchType="exact" searchCase="sensitive" overridternatives="1" always_zap="1">
root@beyonwizv2:~# date --date=@"1632873600"  ##after
Wed Sep 29 08:00:00 AWST 2021
root@beyonwizv2:~# date --date=@"1633046400"  ##before
Fri Oct  1 08:00:00 AWST 2021
root@beyonwizv2:~#

```

**Comment by rdamas** [comment]
What I can see is that only new Modern view is completely wrong: wrong date as you describe (one day off), and meaning of "after" and "before" is reversed.

IMO Classic view is correct; I have old autotimer entries, and everything is displayed like it should according to the entries from autotimer.xml. All entries were edited with OpenWebif's classic view

**Comment by Grumpy-Geoff** [comment]
> IMO Classic view is correct; I have old autotimer entries, and everything is displayed as it should according to the entries in autotimer.xml. All entries were edited with OpenWebif's classic view

That's not what I see and experience.
The classic view is also incorrect, both with its display labeling and with its editing validation.
Open a definition and try and alter the "end" date (labeled as "From").
You'll receive a validation error, stating that it is before the start date.

![AT_date_period_GUI](https://user-images.githubusercontent.com/71379214/134810836-eef169f6-17da-4808-96b9-ecf33aeb85c5.png)

![AT_date_period_file](https://user-images.githubusercontent.com/71379214/134810842-9a018158-2b99-4dfb-8441-9be65b721ca5.png)

Labels are incorrect -
![AT_date_period_OWIF_classic](https://user-images.githubusercontent.com/71379214/134810865-a9401ef9-5779-4eac-a76e-c10a9c1fd5ff.png)

Incorrect date error issued if end date is altered -
![AT_date_period_OWIF_classic#2](https://user-images.githubusercontent.com/71379214/134810900-e323a06b-fa01-44b8-b4da-4d16d431f97d.png)

**Comment by rdamas** [comment]
Yes, this throws a validation error, but the date is accepted and the autotimer is saved correctly. And the corresponding timers will also be created (that's why I mentioned my old autotimer entries - the timers were all created).

**Comment by Grumpy-Geoff** [comment]
> Yes, this throws a validation error, but the date is accepted and the autotimer is saved correctly.

Oh yes you're correct, if I ignore the incorrect validation error and just save the edit then the changed date sticks.
The editing still requires fixing as do the labels.

**Comment by wedebe** [comment]
The AutoTimer labels in the box GUI are weird (_*not* before, *not* after_, wtf?? 🤔🤪)

<img width="1129" alt="Screen Shot 2021-09-26 at 17 47 06" src="https://user-images.githubusercontent.com/9741693/134816690-bcd15cf3-4d88-4f69-9554-bb09ee882a76.png">
 
<img width="589" alt="Screen Shot 2021-09-26 at 17 46 47" src="https://user-images.githubusercontent.com/9741693/134816692-2391383c-9a16-4d32-852b-c08d89461689.png">

<img width="551" alt="Screen Shot 2021-09-26 at 17 43 17" src="https://user-images.githubusercontent.com/9741693/134816696-232bd485-437d-4326-a3a1-eb011345349b.png">

<img width="611" alt="Screen Shot 2021-09-26 at 17 49 10" src="https://user-images.githubusercontent.com/9741693/134816674-90527833-77ce-44ae-8ebf-75e4f51418b2.png">


My entries return the following xml:
[autotimer.xml.txt](https://github.com/E2OpenPlugins/e2openplugin-OpenWebif/files/7231784/autotimer.xml.txt)

**Comment by wedebe** [comment]
Using classic UI:
<img width="560" alt="Screen Shot 2021-09-26 at 18 27 18" src="https://user-images.githubusercontent.com/9741693/134817784-29db2f79-8036-400e-98f9-8b9cd62e1563.png">

**Comment by Grumpy-Geoff** [comment]
> The dates are inclusive, right? So 28th to 30th would match 3 days?

Yes.

> Do timezones affect inputting and/or saving dates?

No, the timeframe is whole days, the plugin adds 24hrs to "end" date (midnight) when checking if EPG event is in range.


> I also wonder whether upcoming summertime changes will cause problems 🙈

We don't have it in Perth ;)  
I've never seen a report of issues regarding timeframe selection on the Beyonwiz forum during DST periods.
The dates are whole days so there's no impact.

**Comment by Grumpy-Geoff** [comment]
`<timer name="GUI 28th to 30th" match="tjgpjmjaj" enabled="yes" id="3" after="1632783603" before="1632956403">`

From 1632783603 to 1632956403
How did you get the non-midnight times in there - did you fudge them?

**Comment by wedebe** [comment]
> `<timer name="GUI 28th to 30th" match="tjgpjmjaj" enabled="yes" id="3" after="1632783603" before="1632956403">`
> 
> From 1632783603 to 1632956403
> How did you get the non-midnight times in there - did you fudge them?

This is what's returned by the /autotimer API - originally entered by using the AutoTimer creator on the box's ui.

**Comment by Grumpy-Geoff** [comment]
> > How did you get the non-midnight times in there - did you fudge them?
> 
> This is what's returned by the /autotimer API - originally entered by using the AutoTimer creator on the box's ui.

Weird.
I just checked my file, and of the 13 definitions that have a date range specified, 3 don't have them as whole days.
I just then created a test definition with a date range via the GUI from the EPG and I observe the same as you.

I think the GUI plugin has an error here, where it appears to me that it nukes the hours and minutes portion but leaves the seconds as they are (I'm not a python programmer) -
https://github.com/oe-alliance/enigma2-plugins/blob/master/autotimer/src/AutoTimerEditor.py#L236

**Comment by wedebe** [comment]
> > > How did you get the non-midnight times in there - did you fudge them?
> > 
> > 
> > This is what's returned by the /autotimer API - originally entered by using the AutoTimer creator on the box's ui.
> 
> Weird. I just checked my file, and of the 13 definitions that have a date range specified, 3 don't have them as whole days. I just then created a test definition with a date range via the GUI from the EPG and I observe the same as you.
> 
> I think the GUI plugin has an error here, where it appears to me that it nukes the hours and minutes portion but leaves the seconds as they are (I'm not a python programmer) - https://github.com/oe-alliance/enigma2-plugins/blob/master/autotimer/src/AutoTimerEditor.py#L236

If only Python had functions to parse and manipulate dates and times 🤣

**Comment by wedebe** [comment]
> Do timezones affect inputting and/or saving dates?
> 
> I also wonder whether upcoming summertime changes will cause problems 🙈

Glad I asked... seems the [developers at Sky didn't consider the hour change](https://www.theregister.com/2021/11/02/sky_plus_clock/) 😄
